File: Helmut/14/solve.jl
function addSand(m)
   q = (500,0)
   k = maximum(xy[2] for xy in keys(m))
   while q[2] <= k 
      if !((q[1],q[2]+1) in keys(m))
          q = (q[1],q[2]+1)
      elseif !((q[1]-1,q[2]+1) in keys(m))
          q = (q[1]-1,q[2]+1)
      elseif !((q[1]+1,q[2]+1) in keys(m))
          q = (q[1]+1,q[2]+1)
      else
          m[q] = 'o'
          return true
      end
   end
   false
end

function bildchen(m)
    println(length(m))
    b = fill(' ', 10, 10)
    for i in 1:10, j in 1:10
        ii = i+494
        jj = j-1
        if (ii,jj) in keys(m) 
             b[j,i] = m[ii,jj]
        end
    end

    for i in 1:10
        println(join(b[i,:]))
    end
end

function read2(xy)
    x,y = split(xy, ",")
    [parse(Int, x), parse(Int, y)]
end

m = Dict()
for l in eachline(open("input.txt"))
    pt = read2.(split(l,"->"))
    q = pt[1]
    for p in pt[2:end]
        dir = sign.(p.-q)
        m[(q[1],q[2])] = '#'
        while q != p 
            q += dir
            m[(q[1],q[2])] = '#'
        end
    end 
end

while addSand(m)
    bildchen(m)
end

println(length([v for v in values(m) if v =='o']))




    

File: Helmut/14/solve.hs
import Data.List.Split (splitOn)
import qualified Data.Set as Set

small = "498,4 -> 498,6 -> 496,6\n503,4 -> 502,4 -> 502,9 -> 494,9"

type P = (Int, Int)
type Grid = Set.Set P

pairs [a,b] = (a,b)

read2 :: String -> (Int, Int)
read2 = pairs . map read . splitOn "," 


add2 (x,y) (u,v) = (x+u, y+v)

connect2points xy@(x,y) uv@(u,v) = let dx = signum (u-x); dy = signum (v-y)
                      in (uv:) $ takeWhile (/=uv) $ iterate (add2 (dx, dy)) xy

connect :: [P] -> [P]
connect ps = concatMap (uncurry connect2points) $ zip ps (tail ps)

filled :: String -> Grid
filled txt = Set.fromList $ concatMap (connect . map read2 . splitOn "->") $ lines txt    

sift :: Grid -> (P, Bool) -> (P, Bool)
sift grid (xy, stuck) = let
                     south = add2 xy (0,1)          
                     sw = add2 xy (-1,1)
                     se = add2 xy (1,1)
                     xy' = head $ dropWhile (\p -> Set.member p grid) [south, sw, se, xy]
                    in (xy', xy'==xy)

addOne :: (Grid, Bool) -> (Grid, Bool)
addOne (grid,_) = let   
                  k = maximum $ Set.map snd grid
                  (xy@(x,y), stuck) = until  (\(xy@(x,y), stuck)-> stuck || y > k) (sift grid) ((500,0), False)
               in if stuck then (Set.insert xy grid, stuck)
                             else (grid, stuck)

main = do 
    grid <- fmap filled (readFile "input.txt")
    let grid' = fst $ until (not.snd) addOne (grid, True)
    print $ length grid' - length grid
